sources/Launcher/DYNCriticalTimeLauncher.cpp

```diff
@@ -0,0 +1,130 @@
+// of simulation tools for power systems.
+//
+
+/**
+ * @file  CriticalTimeLauncher.cpp
+ *
+ * @brief Critical Time launcher: implementation of the algorithm and interaction with dynawo core
+ *
+ */
+
+#ifdef WITH_OPENMP
+#include <omp.h>
+#endif
+
+#include <DYNExecUtils.h>
+#include <DYNSimulation.h>
+#include <DYNSubModel.h>
+#include <DYNModelMulti.h>
+#include "DYNCriticalTimeLauncher.h"
+#include "DYNMultipleJobs.h"
+#include "DYNAggrResXmlExporter.h"
+#include "MacrosMessage.h"
+
+using multipleJobs::MultipleJobs;
+
+namespace DYNAlgorithms {
+
+void
+CriticalTimeLauncher::launch() {
+  const std::string& jobsFile = criticalTimeCalculation_->getJobsFile();
+
+#ifdef WITH_OPENMP
+omp_set_num_threads(nbThreads_);
+#endif
+
+  inputs_.readInputs(workingDirectory_, jobsFile, 1);
+  boost::shared_ptr<job::JobEntry> job = inputs_.cloneJobEntry();
+  SimulationParameters params;
+  boost::shared_ptr<DYN::Simulation> simulation = createAndInitSimulation(workingDirectory_, job, params, results_, inputs_);
+  
+  if (simulation) {
+    boost::shared_ptr<DYN::ModelMulti> modelMulti = boost::dynamic_pointer_cast<DYN::ModelMulti>(simulation->model_);
+    std::string DDBDir = getEnvVar("DYNAWO_DDB_DIR");
+
+    const std::string& dydId = criticalTimeCalculation_->getDydId();
+    const std::string& endPar = criticalTimeCalculation_->getEndPar();
+
+    subModel_ = modelMulti->findSubModelByName(dydId);
+    subModel_->setParameterValue(endPar, DYN::PAR, tSup_, false);
+    subModel_->setSubModelParameters();
+    
+    simulate(simulation, results_);
+  }
+}
+
+void
+CriticalTimeLauncher::updateIndexes(double& tPrevious, double& curAccuracy, const double& multiplierRound) {
+  curAccuracy = std::abs(tSup_ - tPrevious);
+  double midDichotomy = std::round((std::abs(tSup_ - tPrevious) / 2) * multiplierRound) / multiplierRound;
+  double tmp = tSup_;
+
+  if(results_.getSuccess())
+    tSup_ += midDichotomy;
+  else 
+    tSup_ -= midDichotomy;
+
+  tPrevious = tmp;
+}
+
+void
+CriticalTimeLauncher::SearchCriticalTime() {
+  criticalTimeCalculation_ = multipleJobs_->getCriticalTimeCalculation();
+  if (!criticalTimeCalculation_)
+    throw DYNAlgorithmsError(CriticalTimeCalculationTaskNotFound);
+  
+  const double accuracy = criticalTimeCalculation_->getAccuracy(); 
+  double curAccuracy = 1;
+  const double multiplierRound = 1 / accuracy;
+  
+  bool firstSimuPassed = false;
+
+  double tInf = criticalTimeCalculation_->getMinValue();
+  tSup_ = criticalTimeCalculation_->getMaxValue();
+  double tPrevious = tSup_;
+    
+  while(curAccuracy > accuracy) {
+    launch();
+    if (firstSimuPassed)
+      updateIndexes(tPrevious, curAccuracy, multiplierRound);
+    else {
+      if (results_.getSuccess())
+        tSup_ += std::round(((tSup_ - tInf) / 2) * multiplierRound) / multiplierRound;
+      else
+        tSup_ -= std::round(((tSup_ - tInf) / 2) * multiplierRound) / multiplierRound;
```

Comment:
I don't see a stopping condition if tSup goes below tInf

Reply:
The stopping condition is in a certain way the accuracy. If our new value of tSup_ is too close of the previous one, we stop the algorithm. So the tSup value can't goes below tInf.